python/ray/dashboard/memory.py

```diff
@@ -0,0 +1,280 @@
+        else:
+            return False
+
+    def __dict__(self):
+        return {
+            "object_id": self.object_id.hex(),
+            "pid": self.pid,
+            "node_ip_address": self.node_address,
+            "object_size": self.object_size,
+            "reference_type": self.reference_type,
+            "call_site": self.call_site,
+            "local_ref_count": self.local_ref_count,
+            "pinned_in_memory": self.pinned_in_memory,
+            "submittedd_task_ref_count": self.submittedd_task_ref_count,
+            "contained_in_owned": [
+                object_id.hex() for object_id in self.contained_in_owned
+            ],
+            "type": "Driver" if self.is_driver else "Worker"
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MemoryTableSummary:
+    def __init__(self):
+        self.total_object_size = 0
+        self.total_local_ref_count = 0
+        self.total_pinned_in_memory = 0
+        self.total_used_by_pending_task = 0
+        self.total_captured_in_objects = 0
+        self.total_actor_handles = 0
+
+    def summarize(self, table: typing.List[MemoryTableEntry]):
+        for entry in table:
+            if entry.object_size > 0:
+                self.total_object_size += entry.object_size
+            if entry.reference_type == ReferenceType.LOCAL_REFERENCE:
+                self.total_local_ref_count += 1
+            elif entry.reference_type == ReferenceType.PINNED_IN_MEMORY:
+                self.total_pinned_in_memory += 1
+            elif entry.reference_type == ReferenceType.USED_BY_PENDING_TASK:
+                self.total_used_by_pending_task += 1
+            elif entry.reference_type == ReferenceType.CAPTURED_IN_OBJECT:
+                self.total_captured_in_objects += 1
+            elif entry.reference_type == ReferenceType.ACTOR_HANDLE:
+                self.total_actor_handles += 1
+
+    def __dict__(self):
+        return {
+            "total_object_size": self.total_object_size,
+            "total_local_ref_count": self.total_local_ref_count,
+            "total_pinned_in_memory": self.total_pinned_in_memory,
+            "total_used_by_pending_task": self.total_used_by_pending_task,
+            "total_captured_in_objects": self.total_captured_in_objects,
+            "total_actor_handles": self.total_actor_handles
+        }
+
+    def __str__(self):
+        return str(self.__dict__())
+
+
+class MemoryTable:
+    def __init__(self):
+        self.table: typing.List[MemoryTableEntry] = []
+        # Group is a list of memory tables grouped by a group key.
+        self.group: typing.Dict[str, MemoryTable] = self._get_resetted_group()
+        self.summary: MemoryTableSummary = MemoryTableSummary()
+
+    def insert_entry(self, entry: MemoryTableEntry):
+        """Insert a new memory table entry to the table."""
+        self.table.append(entry)
+
+    def setup(self,
```

Comment:
I think we should be calling this in the constructor, not in the `construct_memory_table_function` because it is responsible for summarizing in addition to the sorting and grouping. Before this method is called, the object is not ready to be used, and I think it is ideal to not leave an object in that dangling state unless there's a good reason to.